app/models/app_remote_kill.rb

```diff
@@ -0,0 +1,77 @@
+class AppRemoteKill < ActiveRecord::Base
+  belongs_to :app
+  belongs_to :split
+
+  attribute :first_bad_version, :app_version
+  attribute :fixed_version, :app_version
+
+  validates :app, :split, :reason, :override_to, :first_bad_version, presence: true
+  validates :reason, uniqueness: { scope: %i(app split) }
+
+  validate :override_to_must_exist
+  validate :fixed_version_must_be_greater_than_first_bad_version
+  validate :must_not_overlap_existing
+
+  scope :affecting, ->(app_build, override: false, overridden_at: nil) do
+    where(
+      arel_table[:app_id].eq(app_build.app_id)
+      .and(arel_table[:first_bad_version].lteq(app_build.version))
+      .and(arel_fixed_version_is_null_or_greater_than(app_build.version))
+      .and(
+        Arel::Nodes::Grouping.new(
+          arel_is_false(override)
+          .or(arel_table[:updated_at].gt(overridden_at))
+        )
+      )
+    )
+  end
+
+  class << self
+    private
+
+    def arel_is_false(value_or_arel)
+      Arel::Nodes::Grouping.new(Arel::Nodes::False.new).eq(value_or_arel)
+    end
+  end
+
+  scope :overlapping, ->(other) do
+    where(app_id: other.app_id, split_id: other.split_id)
+      .where.not(id: other.id)
+      .where(
+        if other.fixed_version.nil?
+          Arel::Nodes::True.new
+        else
+          arel_table[:first_bad_version].lt(other.fixed_version)
+        end
+        .and(arel_fixed_version_is_null_or_greater_than(other.first_bad_version))
+      )
+  end
+
+  def self.arel_fixed_version_is_null_or_greater_than(version)
+    Arel::Nodes::Grouping.new(
+      arel_table[:fixed_version].eq(nil)
+      .or(arel_table[:fixed_version].gt(version))
+    )
+  end
+
+  def override_to_must_exist
+    return unless split
+
+    errors.add(:override_to, "must exist in split's current variants") unless split.has_variant?(override_to)
+  end
+
+  def fixed_version_must_be_greater_than_first_bad_version
+    return if fixed_version.nil?
+
+    errors.add(:fixed_version, "must be greater than first bad version") unless fixed_version > first_bad_version
+  end
+
+  def must_not_overlap_existing
+    return if first_bad_version.nil? || split.nil? || app.nil?
+
+    overlapping = self.class.overlapping(self)
```

Comment:
Thoughts on pulling this out?

```ruby
def overlapping
  self.class.overlapping(self)
end
```

Reply:
I guess it only ever applies in an unpersisted context.

Reply:
the validation does run on persisted stuff - there's a where clause to allow IDs to match. I don't feel strongly about it but it's not a mega useful method outside of this context. I'd want to cache it if we put it in an instance var cause i use it twice rapid fire, and I'm not sure about the cache invalidation story then.

Reply:
yeah, come to think about it, i wouldn't be comfortable memoizing  an `#overlapping` method, so I'd still be doing `local_overlapping = overlapping` or something weird here.